### ISSUE TITLE
OpenEBS jiva volume continues to accept reads even after disk loss (simulated) 

### ISSUE THREAD

**Issue by ksatchit** [openebs/openebs#1387]
## Is this a BUG REPORT or FEATURE REQUEST?
BUG REPORT 

**What happened**:

A disk failure/loss was simulated on a single-replica OpenEBS volume by performing a truncate command on the sparse file (volume-head-000.img) in the node where the replica was scheduled. Reads (via dd from the *client* node) on the volume were seen to continue on the iSCSI target - though the Controller logs showed that the iSCSI target was stopped with I/O errors due to loss of backend storage.

**What you expected to happen**:

The truncate command was used to create a sparse file of the same name with a minimal size of 1M, and the read requests were being received for offsets *beyond* the current size of the disk - which should have failed/errored out. However, these continued (possibly returning junk/invalid data)

**How to reproduce it (as minimally and precisely as possible)**:

- Created a volume with a single replica (5G) using a test PVC spec
- Discovered, logged in to iSCSI volume, identified scsi device name & wrote some test data (dd)
- Started continuous reads (dd) w/ bs=1M, count=1000
- Performed truncated file create with same filename as disk file - `truncate -s 1M volume-head-000.img`
- Controller stopped target, changed device mode to RO, threw I/O errors due to *backend not available*

```
time="2018-03-27T18:06:19Z" level=info msg="New Session initiator name:iqn.1993-08.org.debian:01:c4c2da36103e,target name:iqn.2016-09.com.openebs.jiva:pvc-6628f296-31e9-11e8-9309-000c298ff5fc,ISID:0x23d030000"
time="2018-03-27T18:06:19Z" level=error msg="non support"
time="2018-03-27T18:06:19Z" level=warning msg="check condition"
time="2018-03-27T18:06:19Z" level=warning msg="check condition"
time="2018-03-27T18:06:19Z" level=warning msg="check condition"
time="2018-03-27T18:07:14Z" level=error msg="Replicator.ReadAt:0 EOF"
time="2018-03-27T18:07:14Z" level=error msg="Setting replica tcp://10.44.0.5:9502 to ERR due to: EOF"
time="2018-03-27T18:07:14Z" level=info msg="Set replica tcp://10.44.0.5:9502 to mode ERR"
time="2018-03-27T18:07:14Z" level=error msg="I/O error: tcp://10.44.0.5:9502: EOF"
time="2018-03-27T18:07:14Z" level=info msg="stopping target iqn.2016-09.com.openebs.jiva:pvc-6628f296-31e9-11e8-9309-000c298ff5fc ..."
Closing Listener ...
time="2018-03-27T18:07:14Z" level=info msg="target iqn.2016-09.com.openebs.jiva:pvc-6628f296-31e9-11e8-9309-000c298ff5fc stopped"
time="2018-03-27T18:07:14Z" level=info msg="Removing backend: tcp://10.44.0.5:9502"
time="2018-03-27T18:07:14Z" level=info msg="Monitoring stopped tcp://10.44.0.5:9502"
time="2018-03-27T18:07:14Z" level=info msg="Closing: 10.44.0.5:9502"
time="2018-03-27T18:07:14Z" level=error msg="I/O error: No backend available"
time="2018-03-27T18:07:14Z" level=error msg="I/O error: No backend available"
```

**Anything else we need to know?**:

- This test is part of a resiliency suite that aims to inject component failures. The actions in 
this activity are intended to simulate actual disk loss conditions for the OpenEBS jiva volume replica. The
expectation is to fail-fast on such conditions.

**Comment by ksatchit** [comment]
Detaching the external disk upon which the jiva replica is created causes the following logs: 

```time="2019-01-18T12:22:47Z" level=error msg="Error adding replica, err: failed to create temp replica, error: open /openebs/revision.counter: read-only file system, will retry"
time="2019-01-18T12:22:49Z" level=info msg="Closing replica"
time="2019-01-18T12:22:49Z" level=error msg="failed to create temp file: volume.meta while encoding the data to file"
time="2019-01-18T12:22:49Z" level=info msg="Addreplica tcp://10.16.38.64:9502"
time="2019-01-18T12:22:49Z" level=error msg="open existing revision counter file failed"
time="2019-01-18T12:22:49Z" level=error msg="Error in initializing revision counter while creating temp replica"
time="2019-01-18T12:22:49Z" level=error msg="Error adding replica, err: failed to create temp replica, error: open /openebs/revision.counter: read-only file system, will retry"
time="2019-01-18T12:22:51Z" level=info msg="Closing replica"
time="2019-01-18T12:22:51Z" level=error msg="failed to create temp file: volume.meta while encoding the data to file"
time="2019-01-18T12:22:51Z" level=info msg="Addreplica tcp://10.16.38.64:9502"
time="2019-01-18T12:22:51Z" level=error msg="open existing revision counter file failed"
time="2019-01-18T12:22:51Z" level=error msg="Error in initializing revision counter while creating temp replica"
time="2019-01-18T12:22:51Z" level=error msg="Error adding replica, err: failed to create temp replica, error: open /openebs/revision.counter: read-only file system, will retry"
time="2019-01-18T12:22:53Z" level=info msg="Closing replica"
time="2019-01-18T12:22:53Z" level=error msg="failed to create temp file: volume.meta while encoding the data to file"
time="2019-01-18T12:22:53Z" level=info msg="Addreplica tcp://10.16.38.64:9502"
time="2019-01-18T12:22:53Z" level=error msg="open existing revision counter file failed"
time="2019-01-18T12:22:53Z" level=error msg="Error in initializing revision counter while creating temp replica"
time="2019-01-18T12:22:53Z" level=error msg="Error adding replica, err: failed to create temp replica, error: open /openebs/revision.counter: read-only file system, will retry"```

However the replica itself continues to run (which is undesirable) 

Workaround:  A re-attach, remount & subsequent restart of the pod will ensure the data replica resyncs (if quorum is intact). 

Note: Any crashes (or deletes), if they occur will cause the local mount point to be used for this replica.

**Comment by vishnuitta** [comment]
When the disk on which jiva replicas are created gets detached, any subsequent IOs on the disk returns EIO. PR#175 in jiva handles this by doing fatal of the replica, so that, when the disk gets attached, replica starts serving IOs.
This simulation of disk detach, by removing or truncating sparse files, will not help. File descriptor due to open by jiva replica will NOT allow to remove or truncate the sparse file, but, once the replica gets exited, sparse file gets removed.
Hence, closing this issue.

**Comment by vishnuitta** [comment]
re-opening this to cherry-pick the fix to v0.8.x branch and assigning to @utkarshmani1997

**Comment by vishnuitta** [comment]
cherry-pick is also done.. hence, closing
